pkg/puller/downloader.go

```diff
@@ -0,0 +1,173 @@
+package puller
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	d *Downloader
+)
+
+type Downloader struct {
+	NumRetries int
+	ModelDir   string
+}
+
+type Protocol string
+
+const (
+	// Supported Protocols
+	S3    Protocol = "s3://"
+	GS    Protocol = "gs://"
+	PVC   Protocol = "pvc://"
+	File  Protocol = "file://"
+	HTTPS Protocol = "https://"
+)
+
+// TODO: Add more supported protocols
+var SupportedProtocols = []Protocol{S3}
+
+func init() {
+	d = NewDownloader()
+}
+
+func NewDownloader() *Downloader {
+	d := new(Downloader)
+	return d
+}
+
+func SetEnvs(numRetries int, modelDir string) {
+	d.NumRetries = numRetries
+	d.ModelDir = modelDir
+}
+
+func DownloadModel(event EventWrapper) error {
+	modelSpec := event.ModelSpec
+	if modelSpec != nil {
+		modelUri := modelSpec.StorageURI
+		hashString := hash(modelUri)
+		log.Println("Processing:", modelUri, "=", hashString)
+		successFile := filepath.Join(d.ModelDir, fmt.Sprintf("SUCCESS.%d", hashString))
```

Comment:
Should we just use the model file and append `.success` to it as a success indicator? My concern about using a separate success file is that it is more error prone when we update, delete and reload the same model multiple times.
Consider the situation model that the TrainedModel model-1 is deleted and then add back how do we make sure the success file still indicates the right status?

Reply:
It is based on the storageURI. If you upload a different file to the same storageURI, then that is the only thing that breaks the "caching" contract of the SUCCESS file. I am only hashing the storageURI, not the modelName.

Reply:
What is a user create a TrainedModel with different name but use the same storageUri? The real question is what identifies a model. It seems that we are trying to use the storageUri to identity a model which doesn't sound right :(

Reply:
StorageURI is for caching. If a different model name used the same storageURI it should use the already cached model. The only issue in this design is that you can’t update an existing storageURI and expect it to update. Maybe we can set a “hard-override” property to always re-download